static/docs/command-reference/run.md

```diff
@@ -54,23 +54,20 @@ try creating a pipeline.
 
 ### Well-behaved commands
 
-DVC is simple to use, you only need to wrap your commands with `dvc run`, and
-define your dependencies and outputs.
+In a sense, `dvc run` works as a special wrapper for your commands. To prevent
+unexpected behavior when DVC executes them, your commands should ideally follow
+these rules:
 
-However, to prevent unexpected behaviors, ideally, your commands should follow
-some principles:
+- Read exclusively from the specified dependencies.
+- Write exclusively to the specified outputs.
+- Completely rewrite outputs (i.e. do not append or edit).
+- Stop reading and writing files when the command exits.
```

Comment:
> - Read exclusively from the specified dependencies.
> - Write exclusively to the specified outputs.

What if there are optional inputs/outputs that users don't want to mark as deps/outs?

> - Completely rewrite outputs (i.e. do not append

Why? So the next stage has the same results as before? What if the pipeline is meant to be incremental?

> - Stop reading and writing files when the command exits.

How could a command continue writing files after it exits?

Reply:
> What if there are optional inputs/outputs that users don't want to mark as deps/outs?

I think the whole point of the paragraph is to describe an ideal case.

> Why? So the next stage has the same results as before? What if the pipeline is meant to be incremental?

DVC deletes outputs before executing the command again. By depending on the output of the previous run you introduce "cycles" to some extent (command depends on it's own output) to make it reproducible - you need to checkout the previous versions of the output for `dvc repro` to reproduce the current one - it's a mess. That's why we even have an option `--persistent-outs` or something like this - but they are still not documented.

> How could a command continue writing files after it exits?

cc @mroutis @Suor - could you chime in?

Reply:
> I think the whole point of the paragraph is to describe an ideal case.

This :point_up: 

> How could a command continue writing files after it exits?

This is common when you use another tool that keeps running in the background.

Here's more context about it, @jorgeorpinel -- https://github.com/iterative/dvc/issues/2542#issuecomment-546553710

Reply:
> DVC deletes outputs before executing the command again.

Ah, this is an important note we should probably clearly mention then. Added.
Should we remove the `--remove-outs` option from docs and tab completion at this point?

> That's why we even have an option --persistent-outs or something like this - but they are still not documented.

Should we create a separate ticket to document this also? Sounds important.

Reply:
> Should we create a separate ticket to document this also? Sounds important.

We have a ticket but we don't want to execute on it. No one likes this option :) And we've been thinking if there is a better way to solve the problem.

> Should we also remove the `--remove-outs` option from docs and tab completion at this point? 

I hope it's hidden in DVC CLI help. Yes we def should remove it from autocomplete scripts. I would keep in docs for now (with a deprecation note) since users might hit a script where it is used and might wonder how does it behave. (btw, I'm lost a bit - how did we come to `--remove-outs`?)

> Ah, this is an important note we should probably clearly mention then.

Again, I hope it's already mentioned in `dvc repro`\`dvc run`? If not it worth mentioning indeed.

Reply:
> I hope it's hidden in DVC CLI help. Yes we def should remove it from autocomplete scripts.

It's not 🙁 OK I opened iterative/dvc/issues/2988 for this.

> I would keep in docs for now (with a deprecation note) since users might hit a script where it is used and might wonder how does it behave.

OK I reworded it a little bit (in 4778205f) to make the deprecation more obvious.

> btw, I'm lost a bit...

Sorry, I use  "regular update" PRs for misc. small stuff.

> I hope it's already mentioned in dvc repro`dvc run`?

Its currently not, but this PR does introduce the note (in ced036f).